proposals/list-patterns.md

```diff
@@ -0,0 +1,122 @@
+# List patterns
+
+## Summary
+
+Lets you to match an array or a list with a sequence of patterns e.g. `array is {1, 2, 3}` will match an integer array of the length three with 1, 2, 3 as its elements, respectively.
+
+## Detailed Design
+
+The pattern syntax is modified as follow:
+
+```antlr
+recursive_pattern
+  : type? positional_pattern_clause? length_pattern_clause? property_or_list_pattern_clause? simple_designation?
+  ;
+
+length_pattern_clause
+  : '[' pattern ']'
+  ;
+
+property_or_list_pattern_clause
+  : list_pattern_clause
+  | property_pattern_clause
+  ;
+
+property_pattern_clause
+  : '{' (subpattern (',' subpattern)* ','?)? '}'
+  ;
+
+list_pattern_clause
+  : '{' pattern (',' pattern)* ','? '}'
+  ;
+
+slice_pattern
+  : '..' negated_pattern?
+  ;
+
+primary_pattern
+  : recursive_pattern
+  | slice_pattern
+  ;
+```
+There are two new additions to the *recursive_pattern* syntax as well as a *slice_pattern*:
+
+- The *list_pattern_clause* is used to match elements and the *length_pattern_clause* is used to match the length.
+- A *slice_pattern* is only permitted once and only directly in a *list_pattern_clause* and discards _**zero or more**_ elements.
+
+Due to the ambiguity with *property_pattern_clause*, the *list_pattern_clause* cannot be empty and a *length_pattern_clause* should be used instead.
+
+#### Pattern compatibility
+
+A *length_pattern_clause* is compatible with any type that conforms to the following rules:
+
+1. Has an accessible property getter that returns an `int` and has the name `Length` or `Count`
+
+A *list_pattern_clause* is compatible with any type that conforms to the following rules:
+
+1. Is compatible with the length pattern
+2. Has an accessible indexer with a single `int` parameter
```

Comment:
For other recent pattern-based features, we've used a different formulation, instead on relying on equivalent invocation code. This means that an indexer with an optional or `params` parameter would be found acceptable.

Tagging @chsienki to confirm whether I'm remembering right. I wonder if we have any document or example for this guideline.

Reply:
It looks like we may have taken the old approach for range...
Let's add a link to the ranges spec here.
https://github.com/dotnet/csharplang/blob/master/proposals/csharp-8.0/ranges.md#implicit-index-support

Reply:
We have a mixed record on this, as earlier language features deviate in different ways. 

I think in general though there is a desire for new language features to match the same definition of 'pattern' and we have machinery in the compiler now to do so: essentially if it's invocable as a method at the call site then it should satisfy the pattern.

I actually wrote a doc on this, but it never got out of PR. I should probably revive it so we have something to point to going forward.

Reply:
What if the type has an `this[Index]` indexer?

Reply:
>It looks like we may have taken the old approach for range

That's right. I did mention that this is exactly what specified for ranges a few lines below. This was part of the original proposal but afaik never called out in the LDM (presumably ok'd). I'll add a link.